platforms/documentation/docs/src/docs/userguide/optimizing-builds/configuration_cache.adoc

```diff
@@ -536,6 +536,63 @@ And, when an incompatible task is scheduled to run, Gradle discards the configur
 
 Check the method documentation for more details.
 
+[[config_cache:integrity_check]]
+=== Using integrity checks to debug store and load issues
+
+In order to improve performance and reduce entry size, the configuration cache performs a minimal amount of integrity checks as it writes or reads the data.
+However, it makes it harder to troubleshoot issues caused by concurrency problems or misbehaving serialization logic.
+The incorrectly stored object may not be discovered immediately, but it can affect reading data of some other, unrelated task.
+For example, reading a value of the property may end up with a type mismatch or break some other invariant, resulting in a potentially misleading or misattributed error message.
+
+In order to make debugging easier, it is possible to enable stricter checks that may help to discover discrepancies earlier,
+at a cost of slowing down cache operations and significantly increasing the size of the entry.
+To do that, add the following line to your `gradle.properties`:
+[source,properties]
+.gradle.properties
+----
+org.gradle.configuration-cache.integrity-check=true
+----
+
+For example, let's look at a type that implements a custom serialization protocol incorrectly:
+[source,java]
+----
+include::{snippetsPath}/configurationCache/integrityChecks/common/buildSrc/src/main/java/User.java[tags=broken-type]
+----
+<1> `writeObject` serializes both fields.
+<2> `readObject` reads only the first field, leaving the remaining data in the stream.
+
+Such a type will cause problems when used as part of a task state, because the configuration cache will try to interpret the remaining unread part of the object as some new value.
```

Comment:
```suggestion
Such a type will cause problems when used as part of a task state, because the configuration cache will try to interpret the remaining unread part of the object as some new value:

```